src/project.rs

```diff
@@ -0,0 +1,95 @@
+use std::error::Error;
+use std::ffi::OsString;
+use std::fs::{read_dir, DirEntry};
+use std::io::prelude::*;
+
+pub type ProjectError = String;
+pub type Proc<T> = Result<T, ProjectError>;
+
+pub fn discover(dir: &str) -> Proc<Vec<GitRepo>> {
+    match read_dir(dir) {
+        Err(error) => Err(error.description().to_string()),
+        Ok(files) => {
+            for file_result in files {
+                match file_result {
+                    Ok(file) => {
+                        let x = try!(git_repo(file));
+                        println!("Found path: {:?}", x);
+                    }
+                    Err(err) => {
+                        return Err(err.description().to_string());
+                    }
+                }
+            }
+            Err("Finished work".to_string())
+        },
+    }
+}
+
+// TODO: Better parameter type for representing files?
+pub fn git_repo(root: DirEntry) -> Proc<GitRepo> {
```

Comment:
Right. It seems simpler to keep `Option<T>` at the top of the API rather than the more opinionated `Result<T, E>`. The only thing option masks is the _reason_ we might get `None` (that's the type right?) back from scanning a directory. e.g.

1. There was no valid project type (no .git dir, for instance)
2. An internal error occurred